ui-manchette-with-spacetimechart/src/helpers.ts

```diff
@@ -0,0 +1,166 @@
+import type {
+  OperationalPointType,
+  ProjectPathTrainResult,
+  StyledOperationalPointType,
+} from '@osrd-project/ui-manchette/dist/types';
+import type {
+  OperationalPoint,
+  SpaceTimeChartProps,
+} from '@osrd-project/ui-spacetimechart/dist/lib/types';
+
+import { BASE_OP_HEIGHT, MAX_TIME_WINDOW, MAX_ZOOM_X, MIN_ZOOM_X } from './consts';
+import { getHeightWithoutLastWaypoint, msToS } from './utils';
+
+export const calcOperationalPointsToDisplay = (
+  operationalPoints: OperationalPointType[],
+  height: number,
+  isProportional: boolean,
+  zoom: number
+): StyledOperationalPointType[] => {
+  if (!isProportional || operationalPoints.length === 0) {
+    // For non-proportional display, we always display all the operational points:
+    return operationalPoints.map((op) => ({ ...op, styles: {}, display: true }));
+  }
+
+  // For proportional display, we only display points that do not overlap with
+  // the last displayed point:
+  const result: StyledOperationalPointType[] = [
+    { ...operationalPoints[0], display: true, styles: {} },
+  ];
+  const from = operationalPoints[0].position;
+  const to = operationalPoints.at(-1)!.position;
+  const totalDistance = to - from;
+  const heightWithoutFinalOp = getHeightWithoutLastWaypoint(height);
+  let lastDisplayedOP = result[0];
+
+  // We iterate through all points, and only add them if they don't collide
+  // with the last visible point:
+  for (let i = 1; i < operationalPoints.length; i++) {
+    const op = operationalPoints[i];
+    const diff = op.position - lastDisplayedOP.position;
+    const display = (diff / totalDistance) * heightWithoutFinalOp * zoom >= BASE_OP_HEIGHT;
+    result.push({
+      ...op,
+      display,
+      styles: {},
+    });
+
+    if (display) lastDisplayedOP = result[i];
+  }
+
+  // In the end, to make sure the last point is visible, if it's not, we swap
+  // it with the last visible item:
+  const lastItem = result[result.length - 1];
+  if (!lastItem.display) {
+    const lastVisibleItem = result.findLast((op) => op.display);
+    if (lastVisibleItem) {
+      lastVisibleItem.display = false;
+      lastItem.display = true;
+    }
+  }
```

Comment:
If you have filter the non visible points, then you can simplify your code like this:
```suggestion
  if (lastItem.id !== operationalPoints.at(-1)!.id) {
    result.splice(-1, 1, operationalPoints.at(-1)!);
  }
```

Reply:
We can delay the resolution of this comment

Reply:
again thanks !